src/s390_crypto.c

```diff
@@ -675,6 +666,44 @@ int s390_get_functionlist(libica_func_list_element *pmech_list,
 		pmech_list[x].flags = 0;
 		pmech_list[x].property = 0;
 	}
+
+	/* Adjust the flags and properties for algorithms that are allowed in fips
+	 * mode, but not on any hardware, or not with any key length, curve etc. */
+	if (fips & ICA_FIPS_MODE) {
+		/* RSA >= 2048 bits in FIPS 140-3 mode */
+		switch (pmech_list[x].mech_mode_id) {
+		case RSA_KEY_GEN_ME:
+		case RSA_KEY_GEN_CRT:
+		case RSA_ME:
+		case RSA_CRT:
+			if (pmech_list[x].flags)
+				pmech_list[x].property = ICA_PROPERTY_RSA_FIPS;
+			else
+				pmech_list[x].property = 0;
+			break;
+		default:
+			break;
+		}
+
+		/* ECDSA/ECDH only via CPACF in FIPS 140-3 mode */
+		switch (pmech_list[x].mech_mode_id) {
+		case EC_DH:
+		case EC_DSA_SIGN:
+		case EC_DSA_VERIFY:
+		case EC_KGEN:
+			pmech_list[x].flags &= ~ICA_FLAG_DHW;
+			pmech_list[x].flags &= ~ICA_FLAG_SW;
```

Comment:
So here you essentially state that in Fips mode EC operations as well as EC key gen only operate on CPACF (thus no DHW and no SW). 

1. In the next commits you force the EC key gen to happen by OpenSSL, this is SW. So at least for EC Keygen, it thus should be only the SW flag, but no others.
2. Are you actually restricting the EC operations to happen only on CPACF? I don't seem to find that code in any upcoming commit. I only see that you perform selftests only on CPACF (because you need a deterministic K value). But for regular, API driven EC requests it goes the regular path, and thus may choose CPACF, card or SW as available.

So please double check if the flag setting is correct here.